c10/test/util/complex_test_common.h

```diff
@@ -0,0 +1,444 @@
+  static_assert(one_zero<double>().imag() == double(0), "");
+}
+
+} // namespace real_imag
+
+namespace arithmetic_assign {
+
+template<typename scalar_t>
+constexpr c10::complex<scalar_t> p(scalar_t value) {
+  c10::complex<scalar_t> result(scalar_t(2), scalar_t(2));
+  result += value;
+  return result;
+}
+
+template<typename scalar_t>
+constexpr c10::complex<scalar_t> m(scalar_t value) {
+  c10::complex<scalar_t> result(scalar_t(2), scalar_t(2));
+  result -= value;
+  return result;
+}
+
+template<typename scalar_t>
+constexpr c10::complex<scalar_t> t(scalar_t value) {
+  c10::complex<scalar_t> result(scalar_t(2), scalar_t(2));
+  result *= value;
+  return result;
+}
+
+template<typename scalar_t>
+constexpr c10::complex<scalar_t> d(scalar_t value) {
+  c10::complex<scalar_t> result(scalar_t(2), scalar_t(2));
+  result /= value;
+  return result;
+}
+
+template<typename scalar_t>
+C10_HOST_DEVICE void test_arithmetic_assign_scalar() {
+  constexpr c10::complex<scalar_t> x = p(scalar_t(1));
+  static_assert(x.real() == scalar_t(3), "");
+  static_assert(x.imag() == scalar_t(2), "");
+  constexpr c10::complex<scalar_t> y = m(scalar_t(1));
+  static_assert(y.real() == scalar_t(1), "");
+  static_assert(y.imag() == scalar_t(2), "");
+  constexpr c10::complex<scalar_t> z = t(scalar_t(2));
+  static_assert(z.real() == scalar_t(4), "");
+  static_assert(z.imag() == scalar_t(4), "");
+  constexpr c10::complex<scalar_t> t = d(scalar_t(2));
+  static_assert(t.real() == scalar_t(1), "");
+  static_assert(t.imag() == scalar_t(1), "");
+}
+
+template<typename scalar_t, typename rhs_t>
+constexpr c10::complex<scalar_t> p(scalar_t real, scalar_t imag, c10::complex<rhs_t> rhs) {
+  c10::complex<scalar_t> result(real, imag);
+  result += rhs;
+  return result;
+}
+
+template<typename scalar_t, typename rhs_t>
+constexpr c10::complex<scalar_t> m(scalar_t real, scalar_t imag, c10::complex<rhs_t> rhs) {
+  c10::complex<scalar_t> result(real, imag);
+  result -= rhs;
+  return result;
+}
+
+template<typename scalar_t, typename rhs_t>
+constexpr c10::complex<scalar_t> t(scalar_t real, scalar_t imag, c10::complex<rhs_t> rhs) {
+  c10::complex<scalar_t> result(real, imag);
+  result *= rhs;
+  return result;
+}
+
+template<typename scalar_t, typename rhs_t>
+constexpr c10::complex<scalar_t> d(scalar_t real, scalar_t imag, c10::complex<rhs_t> rhs) {
+  c10::complex<scalar_t> result(real, imag);
+  result /= rhs;
+  return result;
+}
+
+template<typename scalar_t>
+C10_HOST_DEVICE void test_arithmetic_assign_complex() {
+  using namespace c10::complex_literals;
+  constexpr c10::complex<scalar_t> x2 = p(scalar_t(2), scalar_t(2), 1.0_if);
+  static_assert(x2.real() == scalar_t(2), "");
+  static_assert(x2.imag() == scalar_t(3), "");
+  constexpr c10::complex<scalar_t> x3 = p(scalar_t(2), scalar_t(2), 1.0_id);
+  static_assert(x3.real() == scalar_t(2), "");
+#if !defined(__CUDACC__)
+  // The following is flaky on nvcc
```

Comment:
This is flaky due to a compiler bug. But this bug seems to only at the evaluation of const expression. When I `std::out` the result, it looks fine.